[fx] split Node into Node/Proxy

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43083 [fx] add type annotations
* #43082 [fx] enabling typechecking of fx files
* **#42991 [fx] split Node into Node/Proxy**

Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

Differential Revision: [D23099786](https://our.internmc.facebook.com/intern/diff/D23099786)

zdevito committed:
[fx] split Node into Node/Proxy

Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

[ghstack-poisoned]

zdevito committed:
Update on "[fx] split Node into Node/Proxy"


Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

[ghstack-poisoned]

zdevito committed:
Update on "[fx] split Node into Node/Proxy"


Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

Differential Revision: [D23099786](https://our.internmc.facebook.com/intern/diff/D23099786)

[ghstack-poisoned]

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -53,13 +65,51 @@ class DefaultDelegate:
+        super().__init__(graph)
+        self.root = root
+
+    def create_arg(self, a):        
```
nit: Trailing whitespace on this line

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -53,13 +65,51 @@ class DefaultDelegate:
+            for n, p in self.root.named_parameters():
+                if a is p:
+                    return self.graph.get_param(n)
+            raise NameError('parameter is not a member of this module')
```
What's going on here? Isn't this covered in the base implementation?

zdevito replied:
Yeah, I refactored wrong in this revision. The fixed version is now more clear: the base delegate cannot handle parameters because it doesn't have a `root` module associated with it. It is used primarily when you are just using the proxy object to add to a graph outside of symbolic tracing, e.g. `t = Proxy(node); output = (t *t - t).node`.

zdevito replied:
The advantage of the "only Node" approach before was that you didn't have to wrap nodes in proxy objects to add more compute with them. But the problems it introduces out-weigh making it slightly more difficult to do.

jamesr66a reviewed (approved):
LGTM, only small comments

zdevito committed:
Update on "[fx] split Node into Node/Proxy"


Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

Differential Revision: [D23099786](https://our.internmc.facebook.com/intern/diff/D23099786)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] split Node into Node/Proxy"


Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

Differential Revision: [D23099786](https://our.internmc.facebook.com/intern/diff/D23099786)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] split Node into Node/Proxy"


Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

Differential Revision: [D23099786](https://our.internmc.facebook.com/intern/diff/D23099786)

[ghstack-poisoned]
